File: script/python/game/service/service_base.py
# -*- encoding:utf-8 -*-

import Game
import Config
from proto.pb_message import Message
from game.util import logger
from game.util.rpc import RpcMgr
from game.util.db_proxy import DBProxy
import game.util.cmd_util

# import asyncio


class ServiceBase:

    rpc_proc = game.util.cmd_util.CmdDispatch("rpc_base")

    def __init__(self, s_cmd, c_cmd, rpc_proc=None):
        self._service_obj = Game.Service()
        self.db_proxy = DBProxy(self)
        self._s_cmd = s_cmd
        self._c_cmd = c_cmd
        if rpc_proc is not None:
            self._rpc_proc = rpc_proc
        else:
            self._rpc_proc = ServiceBase.rpc_proc
        self._rpc_proc.reg_cmd_func("HotFix", ServiceBase._on_recv_hotfix)
        self._rpc_mgr = RpcMgr(self)
        s_cmd.reg_cmd_func(Message.MSG_ID_RPC_MSG_RSP, ServiceBase._on_recv_rpc_rsp_msg)
        s_cmd.reg_cmd_func(Message.MSG_ID_RPC_MSG, ServiceBase._on_recv_rpc_msg)
        print("service obj = ", self._service_obj)

    def on_service_start(self):
        self._init_id_mgr()

    def _init_id_mgr(self):
        redis_ip = Config.getConfigStr("id_redis_ip")
        redis_port = Config.getConfigInt("id_redis_port")
        if not redis_ip or not redis_port:
            return
        import game.util.id_mgr
        game.util.id_mgr.IDMgr.connect_redis(redis_ip, redis_port)

    # def start_asyncio_loop(self):
    #     loop = asyncio.get_event_loop()
    #
    #     # Schedule a call to hello_world()
    #     # loop.call_soon(hello_world, loop)
    #
    #     # Blocking call interrupted by loop.stop()
    #     try:
    #         print("loop start")
    #         loop.run_forever()
    #     finally:
    #         loop.close()
    #         print("loop end")

    # def on_recv_client_msg(self, conn_id, msg_id, msg_data):
    #     logger.logError("$service on_recv_client_msg error, must override in drived class!!!", conn_id, msg_id, msg_data)

    # def on_recv_service_msg(self, sender, msg_id, msg_data):
    #     logger.logError("service on_recv_service_msg error, must override in drived class!!!", sender, msg_id, msg_data)

    def on_recv_client_msg(self, conn_id, msg_id, msg_data):
        logger.log_info("recv client msg, conn_id:{}, msg_id:{}", conn_id, msg_id)
        if self._c_cmd is None:
            logger.log_info("on_recv_client_msg error, not set c_cmd, msgId:{}", msg_id)
            return
        func = self._c_cmd.get_cmd_func(msg_id)
        if func is None:
            logger.log_info("on_recv_client_msg error, not found cmd func, msgId:{}", msg_id)
            return
        msg = Message.create_msg_by_id(msg_id)
        msg.ParseFromString(msg_data)
        func(self, conn_id, msg_id, msg)

    def on_recv_service_msg(self, sender, msg_id, msg_data):
        logger.log_info("recv service msg, sender:{}, msg_id:{}", sender, msg_id)
        func = self._s_cmd.get_cmd_func(msg_id)
        if func is None:
            logger.log_info("on_recv_service_msg error, not found cmd func, msgId:{}", msg_id)
            return
        msg = Message.create_msg_by_id(msg_id)
        msg.ParseFromString(msg_data)
        func(self, sender, msg_id, msg)

    def _on_recv_rpc_msg(self, sender, msg_id, msg):
        if self._rpc_proc is None:
            logger.log_info("on_recv_rpc_msg error, not set rpc proc func, rpc_func:{}", msg.rpc_func)
            return
        func = self._rpc_proc.get_cmd_func(msg.rpc_func)
        if func is None:
            logger.log_info("on_recv_rpc_msg error, not found rpc proc func, rpc_func:{}", msg.rpc_func)
            return
        dict_param = eval(msg.rpc_param)
        rpc_data = func(self, sender, **dict_param)
        rpc_rsp_msg = Message.create_msg_by_id(Message.MSG_ID_RPC_MSG_RSP)
        rpc_rsp_msg.rpc_id = msg.rpc_id
        if rpc_data is not None:
            if type(rpc_data) in (tuple, list):
                rpc_rsp_msg.rpc_data = repr(rpc_data)
            else:
                rpc_rsp_msg.rpc_data = repr((rpc_data,))
        self.send_msg_to_service(sender, rpc_rsp_msg)

    def _on_recv_rpc_rsp_msg(self, sender, msg_id, msg):
        self._rpc_mgr.on_recv_rpc_rsp_msg(sender, msg.rpc_id, msg.rpc_data)

    def _on_recv_hotfix(self, sender):
        import hotfix.hotfix
        hotfix.hotfix.start_hotfix()

    def send_msg_to_client(self, conn_id, msg):
        msg_dat = msg.SerializeToString()
        msg_id = Message.get_msg_id(msg)
        self._service_obj.sendMsgToClient(conn_id, msg_id, msg_dat)

    def send_msg_to_service(self, dst_srv, msg):
        msg_dat = msg.SerializeToString()
        msg_id = Message.get_msg_id(msg)
        self._service_obj.sendMsgToService(dst_srv, msg_id, msg_dat)

    def rpc_call(self, dst_srv, func_name, timeout=RpcMgr.DEFAULT_TIME_OUT, **kwargs):
        return self._rpc_mgr.rpc_call(dst_srv, func_name, timeout, **kwargs)

File: script/lua/service/service_base.lua
require("base.object")
require("util.logger")
require("base.rpc")
require("base.service_type")
require("base.db_proxy")
require("base.id_mgr")
require("util.const")
require("base.hotfix")

clsServiceBase = clsObject:Inherit("clsServiceBase")

function clsServiceBase:__init__()
    self._serviceMsgHandler = {}
    self._clientMsgHandler = {}
    self.db_proxy = clsDBProxy:New(self)
    self._rpc_mgr = clsRpc:New(self)
    self.gm_handler = clsGMHandler:New(self)
    self:_init_id_mgr()
    self:reg_service_msg_handler(MSG_ID_RPC_MSG, self.onRecvRpcMsg)
    self:reg_service_msg_handler(MSG_ID_RPC_MSG_RSP, self.onRecvRpcResp)
    self:reg_rpc_handler("RpcGMCmd", self.rpcGMCmd)
    logger.logDebug("clsServiceBase:__init_")
end

function clsServiceBase:_init_id_mgr()
    local redis_ip = Config:getConfigStr("redis_ip")
    local redis_port = Config:getConfigInt("redis_port")
    if redis_ip == "" and redis_port <= 0 then return end
    IDMgr.connect_redis(redis_ip, redis_port)
end

function clsServiceBase:reg_service_msg_handler(msgId, handler)
    self._serviceMsgHandler[msgId] = handler
end

function clsServiceBase:reg_client_msg_handler(msgId, handler)
    self._clientMsgHandler[msgId] = handler
end

function clsServiceBase:reg_rpc_handler(funcName, func)
    self._rpc_mgr:reg_rpc_handler(funcName, func)
end

function clsServiceBase:on_recv_service_msg(sender, msgId, msg)
    -- print(self:GetType(), sender, msgId, msg, #msg, pb.tohex(msg))
    logger.logInfo("on_recv_service_msg, sender:%s, msgId:%d", sender, msgId)

    local handler = self._serviceMsgHandler[msgId]
    if handler == nil then
        logger.logError("not found service msg handler, msgId:%d", msgId)
        return
    end

    local sender = parseToServiceAddr(sender)

    local pbMsg = decodeMsg(msgId, msg)
    handler(self, sender, msgId, pbMsg)
end

function clsServiceBase:on_recv_client_msg(connId, msgId, msg)
	--if msgId == 15 then msg = string.sub(msg, 1, 6) end
    -- print("on_recv_client_msg, connId:%d, msgId:%d", connId, msgId, msg, #msg, pb.tohex(msg))
    logger.logInfo("on_recv_client_msg, connId:%d, msgId:%d", connId, msgId)
    local handler = self._clientMsgHandler[msgId]
    if handler == nil then
        logger.logError("not found client msg handler, msgId:%d", msgId)
        return
    end
    local pbMsg = decodeMsg(msgId, msg)
    handler(self, connId, msgId, pbMsg)
end

function clsServiceBase:sendMsgToService(dstAddr, msgId, msg)
    logger.logInfo("sendMsgToService to %s, msgId:%d", StrUtil.tableToStr(dstAddr), msgId)
    -- local msgName = MSG_ID_TO_NAME[msgId]
    local data = encodeMsg(msgId, msg)
    -- print(dstAddr, msgId, data, string.len(data))
    Service.sendMsgToService(dstAddr, msgId, data, string.len(data))
end

function clsServiceBase:sendMsgToClient(connId, msgId, msg)
    logger.logInfo("sendMsgToClient to %d, msgId:%d", connId, msgId)
    -- local msgName = MSG_ID_TO_NAME[msgId]
    local data = encodeMsg(msgId, msg)
    -- print(dstAddr, msgId, data, string.len(data))
    Service.sendMsgToClient(connId, msgId, data, string.len(data))
end

function clsServiceBase:broadcastMsgToClient(connIds, msgId, msg)
    if type(msgId) ~= "table" then
        logger.logError("broadcastMsgToClient error, connIds is not table, msgId:%d", msgId)
        return
    end
    logger.logInfo("broadcastMsgToClient to %d, msgId:%d", StrUtil.tableToStr(connIds), msgId)
    -- local msgName = MSG_ID_TO_NAME[msgId]
    local data = encodeMsg(msgId, msg)
    -- print(dstAddr, msgId, data, string.len(data))
    Service.broadcastMsgToClient(connIds, msgId, data, string.len(data))
end

function clsServiceBase:sendMsgToClientKCP(connId, msgId, msg)
    logger.logInfo("sendMsgToClientKCP to %d, msgId:%d", connId, msgId)
    -- local msgName = MSG_ID_TO_NAME[msgId]
    local data = encodeMsg(msgId, msg)
    -- print(dstAddr, msgId, data, string.len(data))
    Service.sendMsgToClientKCP(connId, msgId, data, string.len(data))
end

function clsServiceBase:onRecvRpcMsg(sender, msgId, msg)
    self._rpc_mgr:onRecvRpcMsg(sender, msg)
end

function clsServiceBase:onRecvRpcResp(sender, msgId, msg)
    self._rpc_mgr:onRecvRpcResp(sender, msg)
end

function clsServiceBase:callRpc(dstAddr, funcName, timeout, args)
    return self._rpc_mgr:callRpc(dstAddr, funcName, timeout, args)
end

function clsServiceBase:rpcGMCmd(sender, param)
    local err_code, result = self.gm_handler:handle_gm_cmd(param.cmd, param.params or {})
    return err_code, result
end
